TMEDIA-742 : Medium Promo block

ecarlisle opened the pull request:
## Description

Update Medium Promo Block to Themes v2 architecture.

## Jira Ticket

- [TMEDIA-742 : Medium Promo Block](https://arcpublishing.atlassian.net/browse/TMEDIA-742)

## Acceptance Criteria

- Medium Promo block is migrated to the new architecture.
- Dividing lines between items in the block are included, but any dividing lines framing the block will be done at the layout level

## Test Steps

### Blocks Repo : Storybook, Testing, Linting
1. Checkout this branch `git checkout TMEDIA-742-medium-promo`
2. Start Storybook with `npm run storybook`. Verify stories for Medium Promo.

### Feature Pack

1. Change local feature pack branch with `git checkout TMEDIA-742-medium-promo`
2. Ensure `THEMES_BLOCKS_REPO` in `.env` points to local blocks repo.
3. Start Fusion w/ `npx fusion start -f -l @wpmedia/medium-promo-block`
4. From local PB Editor:
     a. Create new page and add `Medium Promo – Arc Block`.
     b. Set "Display Content Info" in block to `ans-item`, `content-api`, `7TLCY4UZN5F4BA5I4FP62E6NYE`.
     c. Confirm block functionality and cosmetics.


Test with the follow stories
* Video Featured Media - website_url = `/2022/08/31/story-with-video-featured-media/`
* Gallery Featured Media - website_url = `/2022/08/31/story-with-gallery-featured-media/`
* No featured Media - website_url = `/2022/08/26/this-is-a-test-story/`


<img width="1626" alt="Microsoft Edge-2022-08-31-21-32 17" src="https://user-images.githubusercontent.com/868127/187776480-a51faeaf-42f0-4d87-8d23-d5a49eb8181f.png">



## Note
This topic branch was accidentally branched from another topic branch. Please disregard all commits in this topic branch prior to [d26ad32](https://github.com/WPMedia/arc-themes-blocks/pull/1481/commits/d26ad3249db2cab85ff8f06d385376431066f4ec). That all all subsequent commits represent work for this pull request.

## Testing

- [X] Confirmed all the test steps a reviewer will follow above are working.
- [X] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [X] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [X] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [X] Confirmed this PR has unit test files
  - [X] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [X] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [x] Linting code actions have passed.
- [x] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [x] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

Commit:
Fixed no-content default logo

Commit:
Fixed videoInPlace

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
fixed test

Commit:
added fallBackImage to promoImage

Commit:
removed formatURL()

Commit:
code cleanup

nschubach committed:
Recommendations for condition changes and style updates

Commit:
added back formatURL()

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
Merge branch 'tmedia-741-no-content' into tmedia-741-recommendations

# Conflicts:
#	blocks/large-promo-block/features/large-promo/default.jsx

nschubach committed:
Merge pull request #1465 from WPMedia/tmedia-741-recommendations

Recommendations for condition changes and style updates

Commit:
fixed PlayVideoInPlace to use mock-content

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
Tmedia 741 include label (#1468)

* docs: Update type in storybook so that the fallback is story. Mock content is set in storybook and overriden

* fix: Refactor to add back the video and gallery handling shown in storybook

* refactor: Take in data for show flags within presentational component

* docs: Add documentation for fulfilling edge case of promo image lead art

* fix: Update embed markup for video play in place

* fix: Pass down video play in place

* docs: Add todo for image icon

* Continuation of abstraction

* Remove unnecessary props var

* sorting

Co-authored-by: Nick Schubach <nschubach@gmail.com>

Commit:
Merge remote-tracking branch 'origin/arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
Merge remote-tracking branch 'origin/arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
removed unnecessary description, headline and byline author details

Commit:
removed unnecessary comments

Commit:
Update blocks/large-promo-block/features/large-promo/default.test.jsx

Commit:
implementaed localise

Commit:
Merge branch 'tmedia-741-no-content' of github.com:WPMedia/arc-themes-blocks into tmedia-741-no-content

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into tmedia-741-no-content

Commit:
updates gallery-text and video-text in global-phrases block

ecarlisle committed:
Add storybook styles for medium promo block.

ecarlisle committed:
Update medium promo to v2.

ecarlisle committed:
Update tests for medium promo block v2.

ecarlisle committed:
Remove unneeded index file.

ecarlisle committed:
Update promo block package file for v2.

ecarlisle committed:
Add promo block Sass.

ecarlisle committed:
Add mock data for medium promo unit tests.

ecarlisle committed:
Remove unneeded useContent import in unit test file.

ecarlisle committed:
Reconsile accidental merge conflict w/ large block story.

ecarlisle committed:
Update readme

matthewroach commented on .storybook/themes/news.scss:
```diff
@@ -1367,6 +1367,40 @@
+							max-width: 100px,
+						),
+						heading: (
+							font-size: var(--global-font-size-7),
```
We should be using the heading alias tokens

matthewroach commented on .storybook/themes/news.scss:
```diff
@@ -1367,6 +1367,40 @@
+						),
+						paragraph: (
+							display: inline,
+							font-size: var(--global-font-size-4),
```
This should be using the body alias tokens for font and line height

ecarlisle replied:
Updated. In fact, these two properties were not needed.

matthewroach commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -83,20 +129,89 @@ const MediumPromoItem = ({ customFields, arcSite }) => {
+								component={Link}
+								condition={contentUrl}
+								href={formatURL(contentUrl)}
+								onClick={registerSuccessEvent}
```
`Link` component around the Image needs to have `assistiveHidden` prop set to `true`

ecarlisle replied:
Added.

matthewroach commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -83,20 +129,89 @@ const MediumPromoItem = ({ customFields, arcSite }) => {
+	const MediaImage = () =>
+		showImage ? (
+			<Conditional component={Link} condition={contentUrl} href={contentUrl}>
+				<Image
```
We should be using resizer v2 now for rendering images

ecarlisle replied:
@matthewroach : I agree but think we should scope this in another ticket. We need to do this for all eight promo blocks and need a consistent way of getting image `auth` values from ANS story data (perhaps similiar to how we get image URLs with `getImageFromANS`). I'd be happy to create ticket(s) where the AC is to:

1. Create a utility to find `auth` tokens for story images.
2. Implement resize logic for all promo blocks, using the utility to get the `auth` token.

I believe this will help us to stay Agile as resizing isn't specifically cited in [this ticket's AC](https://arcpublishing.atlassian.net/browse/TMEDIA-742).

Please let me know what you think. Thanks!

ecarlisle replied:
@matthewroach : This component now uses Resizer V2 for images coming from ANS. Thank you for including `imageANSToImageSrc` in the components repo!

matthewroach commented on blocks/medium-promo-block/package.json:
```diff
@@ -5,9 +5,10 @@
   "author": "Brent Miller <brent.miller@washpost.com>",
```
```suggestion
  "author": "Arc XP - Themes",
```

matthewroach replied:
Description should also be updated to be "Medium Promo Block"

ecarlisle replied:
Got it. Thanks!

matthewroach reviewed (dismissed):
Along with the comments there is also an issue being shown in Chromatic where there is space left for an image when it shouldn't be, example - https://www.chromatic.com/test?appId=5f91e4721ccaba0022a10782&id=62fe387859e8265ef0ad3ed4

ecarlisle committed:
Update Author name in package file.

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

ecarlisle committed:
Update package description and author.

ecarlisle committed:
Correct Properties:

1. Move imageRatio from imported components to custom fields.
2. Alphabetize custom fields on import.

nschubach commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -83,20 +129,89 @@ const MediumPromoItem = ({ customFields, arcSite }) => {
+			</Conditional>
+		) : null;
+
+	return showHeadline || showImage || showImage || showDescription || showByline || showDate ? (
```
```suggestion
	return showHeadline || showImage || showDescription || showByline || showDate ? (
```
For LGTM message

ecarlisle replied:
Github wouldn't let me update here saying the change was outdated but I did make the change.

ecarlisle committed:
Remove MediaImage function and place logic inside main block logic.

ecarlisle committed:
Remove redundant conditional.

ecarlisle committed:
Fix merge conflict.

nschubach commented on .storybook/themes/news.scss:
```diff
@@ -1367,6 +1367,40 @@
+						),
+						paragraph: (
+							display: inline,
+							font-size: var(--global-font-size-4),
+							line-height: var(--global-line-height-4),
```
```suggestion
```
The default paragraph already uses the `body-font-size` and `body-font-line-height` which are `font-size-4` and the appropriate line-height for the font so these aren't needed unless you are overriding the Paragraph default.

ecarlisle replied:
Exactly! Again, committed before I saw this feedback. These lines have been removed.

nschubach commented on .storybook/themes/news.scss:
```diff
@@ -1367,6 +1367,40 @@
+							max-width: 100px,
+						),
+						heading: (
+							font-size: var(--global-font-size-7),
+							font-weight: var(--global-font-weight-7),
+							line-height: var(--global-line-height-4),
```
```suggestion
							font-size: var(--heading-level-5-font-size),
							font-weight: var(--heading-level-5-font-weight),
							line-height: var(--heading-level-5-line-height),
```

ecarlisle replied:
Thanks! Changes have been made! Committed before I saw this commit suggestion.

nschubach commented on .storybook/themes/news.scss:
```diff
@@ -1895,6 +1929,39 @@
+							margin: 0,
+							max-width: 33%,
+						),
+						heading: (
+							font-size: var(--global-font-size-11),
+							line-height: var(--global-line-height-7),
+						),
```
```suggestion
```
These aren't needed with the heading-level-5 changes above.

ecarlisle replied:
Sorry, got these but must not have pushed up the changes. 
These have been removed.

ecarlisle committed:
Use global tokens for heading & other corrections.

ecarlisle committed:
Set Promo blocks paragraphs to inline display at default breakpoint.

ecarlisle committed:
Remove redundant check for rendering block.

ecarlisle commented:
> Along with the comments there is also an issue being shown in Chromatic where there is space left for an image when it shouldn't be, example - https://www.chromatic.com/test?appId=5f91e4721ccaba0022a10782&id=62fe387859e8265ef0ad3ed4

Thank you, @matthewroach ! I've updated the CSS so the image will float to the right on the default viewport range. I've confirmed the change in Chromatic as well.

matthewroach commented:
@ecarlisle - Chromatic is still showing regression with the description not aligned correctly on desktop

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.1' into TMEDIA-742-medium-promo

ecarlisle committed:
Update mock data to account for image resizing.

ecarlisle committed:
Add logic for image resizes on an ANS image.

ecarlisle committed:
Update storybook environment file.

ecarlisle committed:
Update Medium Promo mock data for storybook.

ecarlisle committed:
Use imageANSToImageSrc to get image filename.

ecarlisle committed:
Update package lock file.

ecarlisle committed:
Fix spacing issue with Medium Promo paragraphs at desktop size.

ecarlisle commented:
> @ecarlisle - Chromatic is still showing regression with the description not aligned correctly on desktop

@matthewroach : I believe I've addressed the issue you spoke of. There are several Chromatic changes now, but these appear to be related to lower-level styling changes. I did need to update my `package-lock.json` in this PR so that I could use the `imageANSToImageSrc` function from the components repo.

Please let me know if there are any other Chromatic issues I should address.

JackHowa commented on blocks/medium-promo-block/README.md:
```diff
@@ -1,52 +1,51 @@
 # `@wpmedia/medium-promo-block`
 
-_Block containing a medium promo component. Please provide a 1-2 sentence description of what the block is and what it does._
+Medium Manual Promo Block provides a way for a PageBuilder user to create a data-driven promo item using content data.
```
```suggestion
Medium Promo Block provides a way for a PageBuilder user to create a data-driven promo item using content data.
```

this is not the manual promo

JackHowa replied:
@ecarlisle it looks like this has not been addressed. When you set "ready for review", our team expectation is that all reviews have either been addressed or commented on. I changed this back to "needs changes"

ecarlisle replied:
TYVM @JackHowa ! Corrected.

JackHowa replied:
updated -- sorry to post in slack to remind you on this

JackHowa reviewed (dismissed):
In slack, we talked about checking the chromatic ahead of submitting this for pr. is this the change that's expected?

It looks like the headline only view doesn't account for when an image is not opted into with showImage. if you look in the medium manual promo, this is handled. 

https://www.chromatic.com/test?appId=5f91e4721ccaba0022a10782&id=630cfb4ca02ebcdb8b4ba98c

<img width="717" alt="Screen Shot 2022-08-29 at 13 28 46" src="https://user-images.githubusercontent.com/5950956/187272394-85caa8fe-b567-466a-bcb6-901a17a08197.png">

ecarlisle commented:
> In slack, we talked about checking the chromatic ahead of submitting this for pr. is this the change that's expected?
> 
> It looks like the headline only view doesn't account for when an image is not opted into with showImage. if you look in the medium manual promo, this is handled.
> 
> https://www.chromatic.com/test?appId=5f91e4721ccaba0022a10782&id=630cfb4ca02ebcdb8b4ba98c
> 
> <img alt="Screen Shot 2022-08-29 at 13 28 46" width="717" src="https://user-images.githubusercontent.com/5950956/187272394-85caa8fe-b567-466a-bcb6-901a17a08197.png">

@JackHowa : Thanks! Noticed this too, unfortunately after upload. Will be sure to verify locally before requesting review. Thanks!

ecarlisle committed:
Use custom field showImage to control respective CSS class.

JackHowa commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -121,32 +250,35 @@ MediumPromo.propTypes = {
 		}),
-		...imageRatioCustomField("imageRatio", "Art", "16:9"),
+		imageRatio: PropTypes.oneOf(["16:9", "3:2", "4:3"]).tag({
+			defaultValue: "3:2",
```
Was 3:2 the default aspect ratio before? in the mocks it look like the default is 4:3

<img width="845" alt="Screen Shot 2022-08-29 at 14 22 29" src="https://user-images.githubusercontent.com/5950956/187281556-3d14d8ac-6c1a-4e25-8171-a04c5a734b5d.png">

matthewroach replied:
Default was `16:9` the last argument of the function is the default value

JackHowa replied:
yep

ecarlisle replied:
@matthewroach , @JackHowa : Thank you! I've changed the default to `16:9`.

ecarlisle commented:
@matthewroach , @JackHowa : In all of the Chromatic changes, the biggest difference I'm seeing is the view at 678px. In v1 the image is 100px wide and in v2 we're using a max-width of 33%.

<img width="1917" alt="image" src="https://user-images.githubusercontent.com/452134/187290718-2ba0352d-afef-4d13-b3cd-beebc737a66c.png">

In version 1, though, the change in image size happens when the viewport is 1px above when the layout change happens. 

## Example

### 767px wide
<img width="773" alt="image" src="https://user-images.githubusercontent.com/452134/187294751-07dcbf73-926e-4ebc-8acc-46cf0ee4c9b4.png">

### 768px wide
<img width="766" alt="image" src="https://user-images.githubusercontent.com/452134/187294863-7d382ac9-5449-425e-965d-8f69637d204c.png">

### 768px wide
<img width="763" alt="image" src="https://user-images.githubusercontent.com/452134/187295256-a5cfb7dc-2665-453d-ac3f-484c090a0383.png">

While the V2 PR isn't 100% parity for 768px width, I think it may be a layout improvement worth keeping. Thoughts? Should this be either a Product or VQA call?

Thanks!
Eric

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.1' into TMEDIA-742-medium-promo

ecarlisle committed:
Change default value of imageRatio custom field to 16:9.

ecarlisle committed:
Merge branch 'TMEDIA-742-medium-promo' of github.com:WPMedia/arc-themes-blocks into TMEDIA-742-medium-promo

matthewroach commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -63,6 +113,10 @@ const MediumPromoItem = ({ customFields, arcSite }) => {
+					}
           type
           url
           resized_params {
```
We can remove all references to `resized_params` in the filter as these are no longer used anywhere

ecarlisle replied:
@matthewroach : Thanks! Removed.

matthewroach commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -46,6 +92,10 @@ const MediumPromoItem = ({ customFields, arcSite }) => {
         type
         url
         lead_art {
+					_id
```
Spacing looks off here, is this a tabs vs spaces difference?

ecarlisle replied:
@matthewroach: Yes, thanks! Updated. The query was in spaces and my updates were tabbed. I updated so that the entire filter uses tabs for indention.

ecarlisle committed:
Fix indention on content source filter.

ecarlisle committed:
Remove resized_params from content source filter.

matthewroach commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -20,83 +66,150 @@ const MediumPromoItem = ({ customFields, arcSite }) => {
+					}
+					type
+					promo_items {
+						basic {
```
Is `promo_items` ever used? If so, we need to add `auth` and `_id` here otherwise it will break

ecarlisle replied:
@matthewroach : This one is for `[content].lead_art.promo_items`. So far I haven't had a use case where `auth` or `_id` are needed here, but let me know if I'm overlooking. I've the `_id` and `auth` included for:

1. `[content].lead_art`
2.`[content].promo_items.basic`

Those seem to be the two cases we're covering for a story's Featured Media.

matthewroach replied:
Correct that's the location - Does the block ever need that content? If it's not using it we can remove these items from the `filter` but are we 100% sure the block isn't supposed to be using it? Are we introducing a regression? or are we missing a test case?

ecarlisle replied:
This block doesn't use these fields. I will remove and test to ensure I'm not introducing a regression, also checking v1 of the block.

matthewroach replied:
It was needed - I've updated the filter to match the different items needed

ecarlisle committed:
Change block name in readme.

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.1' into TMEDIA-742-medium-promo

ecarlisle committed:
Update resizer logic.

matthewroach committed:
Account for no authors

matthewroach committed:
Update arc themes components

matthewroach committed:
Add missing label to images

Fix images not showing for different story types

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.1' into TMEDIA-742-medium-promo

matthewroach committed:
Remove resized_params

ecarlisle commented:
@matthewroach : I have verified expected functionality locally and have compared the diffs in Chromatic. I'm satisfied that we've met acceptance criteria. As per your comment above, should I label this PR with "Additional Review" or were you asking me to reverify?
Thank you very much for your help!

Here are local examples I'm seeing as per the PR cases:
![large-promo-block-use-cases](https://user-images.githubusercontent.com/452134/187922895-2e2772e6-f9d7-4db4-94de-eb018da18d5b.png)

nschubach commented:
Change base branch to 2.0.2

ecarlisle committed:
Update package.lock by re-running
> arc-themes-blocks@ postinstall /Users/carlislee/repos/arc-themes-blocks
> npx lerna clean -y

> arc-themes-blocks@ prepare /Users/carlislee/repos/arc-themes-blocks
> husky install

husky - Git hooks installed
added 6 packages from 4 contributors, updated 1 package and audited 4263 packages in 19.89s

347 packages are looking for funding
  run `npm fund` for details

found 258 vulnerabilities (1 low, 72 moderate, 140 high, 45 critical)
  run `npm audit fix` to fix them, or `npm audit` for details after merging with base branch.

malavikakoppula reviewed (approved):
LGTM. Tested in Storybook desktop and mobile versions, Tested in PB both desktop and mobile, also tested no-content fallback image.

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.2' into TMEDIA-742-medium-promo

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.2' into TMEDIA-742-medium-promo

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.2' into TMEDIA-742-medium-promo

ecarlisle committed:
Merge with arc-themes-release-version-2.0.2, update package file.

JackHowa commented:
cool looks like the changes addressed
